### PR TITLE
docs(eslint-plugin): add warning about superfluous rules with typescript

### LINKED ISSUE #7328: Docs: Document Superfluous Rules with TypeScript
### Before You File a Documentation Request Please Confirm You Have Done The Following...

- [X] I have looked for existing [open or closed documentation requests](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Adocumentation) that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

#### @typescript-eslint/no-dupe-class-members

The [ESLint documentation for the rule](https://eslint.org/docs/latest/rules/no-dupe-class-members#when-not-to-use-it) says that the rule is superfluous if you use TypeScript.

According to Brad, the only reason that `typescript-eslint` bothers to maintain an updated version of the rule is because some people prefer the ESLint errors over the TypeScript compiler errors. However, this reasoning is not documented anywhere, so I propose the following warning be added to [the top of the page for `no-dupe-class-members`](https://typescript-eslint.io/rules/no-dupe-class-members):

> Danger
The code problem checked by this ESLint rule is automatically checked by the TypeScript compiler. Thus, it is not recommended to turn on this rule in new TypeScript projects. You only need to enable this rule if you prefer the ESLint error messages over the TypeScript compiler error messages. (When this rule was originally written, the TypeScript compiler checks did not exist.)

#### @typescript-eslint/no-invalid-this

The [ESLint documentation for the rule](https://eslint.org/docs/latest/rules/no-invalid-this#when-not-to-use-it) does not explicitly mention it, but this rule is superfluous if you use TypeScript with the `strict` or `noImplicitThis` compiler flags enabled. (The `strict` flag automatically includes `noImplicitThis`.)

According to Brad, the only reason that `typescript-eslint` bothers to maintain an updated version of the rule is because some people prefer the ESLint errors over the TypeScript compiler errors. However, this reasoning is not documented anywhere, so I propose the following warning be added to [the top of the page for `no-invalid-this`](https://typescript-eslint.io/rules/no-invalid-this/) (which is a slightly modified version of the warning from the previous section):

> Danger
The code problem checked by this ESLint rule is automatically checked by the TypeScript compiler (as long as you have the `strict` or `noImplicitThis` compiler flags enabled). Thus, it is not recommended to turn on this rule in new TypeScript projects. You only need to enable this rule if you prefer the ESLint error messages over the TypeScript compiler error messages. (When this rule was originally written, the TypeScript compiler checks did not exist.)

### PR DESCRIPTION
**Pull request by Zamiell**
## PR Checklist

- [X] Addresses an existing open issue: fixes #7328
- [X] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [X] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Adding warning to docs. We probably want to turn this into a template instead, but not sure how to go about that right now.

### CODE REVIEW

File: packages/eslint-plugin/docs/rules/no-dupe-class-members.md [hunk 1]
```diff
@@ -6,5 +6,11 @@ description: 'Disallow duplicate class members.'
 >
 > See **https://typescript-eslint.io/rules/no-dupe-class-members** for documentation.
 
+:::danger
+
+The code problem checked by this ESLint rule is automatically checked by the TypeScript compiler. Thus, it is not recommended to turn on this rule in new TypeScript projects. You only need to enable this rule if you prefer the ESLint error messages over the TypeScript compiler error messages. (When this rule was originally written, the TypeScript compiler checks did not exist.)
+
+:::
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/docs/rules/no-dupe-class-members.md:13]: Yeah I'm not a huge fan of copypasta. At the very least, let's add a shared JSX component that the rules can embed.

It'd be even more automagical to add it to the rule metadata. Maybe a `supercededByTypeScript` property in `RuleMetaDataDocs`? cc @bradzacher & @Josh-Cena from #4156.

**Review by JoshuaKGoldberg — APPROVED** [review]
LGTM (thanks!), but will wait for other input since maybe we'd want to automate this a bit more?

File: packages/eslint-plugin/docs/rules/no-dupe-class-members.md (hunk 1 above)
Comment (Josh-Cena) [reply packages/eslint-plugin/docs/rules/no-dupe-class-members.md:13]: I would suggest [imported Markdown](https://docusaurus.io/docs/markdown-features/react#importing-markdown) in this case

Comment (Zamiell) [reply packages/eslint-plugin/docs/rules/no-dupe-class-members.md:13]: > But TBH if it's just two rules

There's more than 2 rules, I can add the other rules that this applies to in a future PR

Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/docs/rules/no-dupe-class-members.md:13]: Ooh what are the other rules?

File: packages/eslint-plugin/docs/rules/no-dupe-class-members.md [hunk 2]
```diff
@@ -6,5 +6,11 @@ description: 'Disallow duplicate class members.'
 >
 > See **https://typescript-eslint.io/rules/no-dupe-class-members** for documentation.
 
+:::danger
+
+The code problem checked by this ESLint rule is automatically checked by the TypeScript compiler. Thus, it is not recommended to turn on this rule in new TypeScript projects. You only need to enable this rule if you prefer the ESLint error messages over the TypeScript compiler error messages. (When this rule was originally written, the TypeScript compiler checks did not exist.)
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/no-dupe-class-members.md:11]:
```suggestion
The code problem checked by this ESLint rule is automatically checked by the TypeScript compiler. Thus, it is not recommended to turn on this rule in new TypeScript projects. You only need to enable this rule if you prefer the ESLint error messages over the TypeScript compiler error messages.
```

these checks always existed - the eslint rule just also existed and people liked it!

File: packages/eslint-plugin/docs/rules/no-invalid-this.md [hunk 3]
```diff
@@ -6,5 +6,11 @@ description: 'Disallow `this` keywords outside of classes or class-like objects.
 >
 > See **https://typescript-eslint.io/rules/no-invalid-this** for documentation.
 
+:::danger
+
+The code problem checked by this ESLint rule is automatically checked by the TypeScript compiler (as long as you have the `strict` or `noImplicitThis` compiler flags enabled). Thus, it is not recommended to turn on this rule in new TypeScript projects. You only need to enable this rule if you prefer the ESLint error messages over the TypeScript compiler error messages. (When this rule was originally written, the TypeScript compiler checks did not exist.)
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/no-invalid-this.md:11]:
```suggestion
The code problem checked by this ESLint rule is automatically checked by the TypeScript compiler (as long as you have the `strict` or `noImplicitThis` compiler flags enabled). Thus, it is not recommended to turn on this rule in new TypeScript projects. You only need to enable this rule if you prefer the ESLint error messages over the TypeScript compiler error messages.
```

ditto above

**Review by bradzacher — DISMISSED** [review]

File: packages/eslint-plugin/docs/rules/no-dupe-class-members.md (hunk 1 above)
Comment (Zamiell) [reply packages/eslint-plugin/docs/rules/no-dupe-class-members.md:13]: everything here: https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/configs/eslint-recommended.ts

Comment (Zamiell) [reply packages/eslint-plugin/docs/rules/no-dupe-class-members.md:13]: notice that the two warnings that I added in this PR are slightly different, so any proposed templating mechanism to reduce DRY would need to be able to take parameters.

@Josh-Cena does imported markdown support that?

**Comment by Zamiell** [comment]
I haven't heard back from @Josh-Cena in 2 weeks, so I'll disregard his comment and try using a JSX component.

**Comment by Zamiell** [comment]
Ok, things are looking good now with an MDX component.

I also performed a short audit of the rules and found that the warning message also needed to be added to the `no-redeclare` rule.

This PR is now ready for review again.

File: packages/eslint-plugin/docs/rules/no-dupe-class-members.md (hunk 1 above)
Comment (Zamiell) [reply packages/eslint-plugin/docs/rules/no-dupe-class-members.md]: For now, I removed the duplication with a shared JSX component.

**Comment by Josh-Cena** [comment]
Oh sorry, I didn't realize there are things pending on my feedback. FYI @Zamiell MDX partials also support taking props:

https://docusaurus.io/docs/markdown-features/react#importing-markdown

It should be as easy as taking your component's returned JSX element and converting it to Markdown.

I don't have much objection to the current approach, though.

**Comment by Zamiell** [comment]
Looks like the "CI / Website tests (pull_request)" is failing, but I don't think it has anything to do with my PR, correct me if I am wrong.

**Comment by Zamiell** [comment]
It's been 3 weeks and I haven't heard anything from anyone on the team, so pinging Josh/Brad.

**Review by JoshuaKGoldberg — APPROVED** [review]
✨ super, thanks!

I like this usage of mdx 😄

**Pull request merged**